### PR TITLE
Promote GH gauge constraint from test to function

### PR DESCRIPTION
**Pull request by geoffrey4444**
## Proposed changes

Move GH gauge constraint from a test file (VerifyGrSolution.hpp) to GeneralizedHarmonic/Constraints.?pp, and test by checking for Kerr-Schild and with fuzzy testing.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [x] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [x] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [x] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp
Comment (geoffrey4444) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: Originally, Mark's code checked that the solution was static. That's not necessary to test the gauge constraint, right? But these checks will pass, because I'm only checking Kerr Schild.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp
Comment (osheamonn) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp:70]: Is this version of the function tested?

Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp:70]: Unless I screwed something up. The check_with_random_values test calls this one (return by value), which then calls the other one (return by pointer), thus testing both. Or at a least, that's what's supposed to happen...did I screw up somewhere?

Comment (osheamonn) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp:70]: I'd like to see this function explicitly tested though, since the fact that one calls the other is an implementation detail. Since you've two tests, the analytic and the random version, I'd be fine with just adding a call to this version into the analytic test, and checking that it also gives zero.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp:55]: Why does the one-form have an up index?

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: the solution can't be const?

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: can you set this to `0.0` please?

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: can't this be `const`?

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: Why do you need to do this if the allocation is already set? Doing this allocation completely defeats the purpose of the return-by-reference approach. You should check if the allocation is necessary and only do it if it is.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp]: Can you make the pointer `const`?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp
Comment (geoffrey4444) [reply tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: I can, but isn’t a nonzero choice better in general?

Comment (nilsdeppe) [reply tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Constraints.cpp]: I don't see how it's better if it won't matter. `signalingNaN` would be better, though.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp
Comment (hrueter) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp:55]: @nilsdeppe I guess he means a vector field and not a 1-form.
It think it is sensible to use the vector field here.

Comment (geoffrey4444) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp:55]: Typo; I meant to say vector

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp
Comment (nilsdeppe) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.hpp]: maybe `vector` instead of `vector form`?

**Review by nilsdeppe — COMMENTED** [review]
I'm happy for you to squash when you fix my last small comment :)

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
@wthrowe @kidder @osheamonn @markscheel could you please look at this whenever you get time? Thanks! :)

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@wthrowe please look at

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:81]: can this save operations by computing $\pi_{bc} g^{bc}$ first in a loop, then subtracting from the constraint outside the double-loop? (if so, same on line 90)

**Review by moxcodes — COMMENTED** [review]
Just a quick question - if it's not worth it at this stage, no worries.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (wthrowe) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:81]: @nilsdeppe Thoughts?

**Review by wthrowe — APPROVED** [review]
I'm fine with this, but I'll let Nils take a look at the last comment and see if if he wants a change.

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (nilsdeppe) [reply src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:81]: I think this would add an allocation, so it's not obvious to me it'll be faster and would need benchmarking. I'd be fine with @geoffrey4444 filing an issue and coming back to it later :)

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:70]: `constraint->get(a)` and many times below (arrow operator serves to dereference and access).

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:83]: can avoid `if` statement by doing this work in a separate loop `for(size_t a=1;...)`

File: src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp
Comment (moxcodes) [inline src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp:79]: because all expressions summed over are symmetric in b and c, you can start c at b+1 (with appropriate factors of 2 in each expression) and add the diagonal term afterward for savings.

**Review by moxcodes — COMMENTED** [review]
Some thoughts for possible optimization if/when desired. (code review with @markscheel @fmahebert @mgiesler)

**Comment by geoffrey4444** [comment]
Should I create an issue to look into the suggested optimizations? Or would you like me to do any of these before we push the merge button? I'm fine either way.

**Comment by kidder** [comment]
you can create an issue

**Pull request merged**
